src/compiler/program.ts

```diff
@@ -176,6 +176,39 @@ namespace ts {
         return undefined;
     }
 
+   function tryReadMainSection(packageJsonPath: string, baseDirectory: string, state: ModuleResolutionState): string {

```

Comment:
Nevermind. I see that this is necessary to resolve modules relative to the `tsconfig.json`.

There's a lot of similar logic between `tryReadMainSection` and `tryReadTypesSection`. Can we split out shared logic to something more reusable?